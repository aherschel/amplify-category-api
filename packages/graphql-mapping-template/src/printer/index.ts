export { jsPrinter } from './js-printer';
export { vtlPrinter } from './vtl-printer';
export { Printer } from './printer';
