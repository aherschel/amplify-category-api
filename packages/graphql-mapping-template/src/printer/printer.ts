import {
  Expression,
  IfNode,
  IfElseNode,
  AndNode,
  OrNode,
  ParensNode,
  EqualsNode,
  NotEqualsNode,
  ForEachNode,
  StringNode,
  IntNode,
  NullNode,
  ReferenceNode,
  QuietReferenceNode,
  ObjectNode,
  ListNode,
  FloatNode,
  QuotesNode,
  RawNode,
  SetNode,
  CompoundExpressionNode,
  CommentNode,
  ToJsonNode,
  BooleanNode,
  compoundExpression,
  comment,
  NotNode,
  NewLineNode,
  ReturnNode,
  IsNullOrEmptyNode,
} from '../ast';

export abstract class Printer {
  // Abstract methods implemented by specific printers.
  protected abstract printIf(node: IfNode, indent: string);
  protected abstract printIfElse(node: IfElseNode, indent: string);
  protected abstract printAnd(node: AndNode, indent: string): string;
  protected abstract printOr(node: OrNode, indent: string): string;
  protected abstract printParens(node: ParensNode, indent: string): string;
  protected abstract printEquals(node: EqualsNode, indent: string): string;
  protected abstract printNotEquals(node: NotEqualsNode, indent: string): string;
  protected abstract printForEach(node: ForEachNode, indent: string): string;
  protected abstract printString(node: StringNode): string;
  protected abstract printBool(node: BooleanNode): string;
  protected abstract printRaw(node: RawNode, indent: string): string;
  protected abstract printQuotes(node: QuotesNode): string;
  protected abstract printInt(node: IntNode): string;
  protected abstract printFloat(node: FloatNode): string;
  protected abstract printNull(node: NullNode): string;
  protected abstract printReference(node: ReferenceNode): string;
  protected abstract printQuietReference(node: QuietReferenceNode, indent: string): string;
  abstract printObject(node: ObjectNode, indent: string): string;
  protected abstract printList(node: ListNode, indent: string): string;
  protected abstract printSet(node: SetNode, indent: string): string;
  protected abstract printComment(node: CommentNode, indent: string): string;
  protected abstract printCompoundExpression(node: CompoundExpressionNode, indent: string): string;
  protected abstract printToJson(node: ToJsonNode, indent: string): string;
  protected abstract printIsNullOrEmpty(node: IsNullOrEmptyNode, indent: string): string;
  protected abstract printNot(node: NotNode, indent: string): string;
  protected abstract printNewLine(node: NewLineNode): string;
  protected abstract printReturn(node: ReturnNode, indent: string): string;

  protected printExpr(expr: Expression, indent: string = ''): string {
    if (!expr) {
      return '';
    }
    switch (expr.kind) {
      case 'If':
        return this.printIf(expr, indent);
      case 'IfElse':
        return this.printIfElse(expr, indent);
      case 'And':
        return this.printAnd(expr, indent);
      case 'Or':
        return this.printOr(expr, indent);
      case 'Parens':
        return this.printParens(expr, indent);
      case 'Equals':
        return this.printEquals(expr, indent);
      case 'NotEquals':
        return this.printNotEquals(expr, indent);
      case 'ForEach':
        return this.printForEach(expr, indent);
      case 'String':
        return this.printString(expr);
      case 'Raw':
        return this.printRaw(expr, indent);
      case 'Quotes':
        return this.printQuotes(expr);
      case 'Float':
        return this.printFloat(expr);
      case 'Int':
        return this.printInt(expr);
      case 'Boolean':
        return this.printBool(expr);
      case 'Null':
        return this.printNull(expr);
      case 'Reference':
        return this.printReference(expr);
      case 'QuietReference':
        return this.printQuietReference(expr, indent);
      case 'Object':
        return this.printObject(expr, indent);
      case 'List':
        return this.printList(expr, indent);
      case 'Set':
        return this.printSet(expr, indent);
      case 'Comment':
        return this.printComment(expr, indent);
      case 'CompoundExpression':
        return this.printCompoundExpression(expr, indent);
      case 'Util.ToJson':
        return this.printToJson(expr, indent);
      case 'Util.isNullOrEmpty':
        return this.printIsNullOrEmpty(expr, indent);
      case 'Not':
        return this.printNot(expr, indent);
      case 'NewLine':
        return this.printNewLine(expr);
      case 'Return':
        return this.printReturn(expr, indent);
      default:
        return '';
    }
  }

  print(expr: Expression): string {
    return this.printExpr(expr);
  }

  printBlock(name: string) {
    return (expr: Expression): string => {
      return this.printExpr(compoundExpression([comment(`[Start] ${name}.`), expr, comment(`[End] ${name}.`)]));
    };
  }
}
