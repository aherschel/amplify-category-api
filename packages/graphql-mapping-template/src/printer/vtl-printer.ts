import { Printer } from './printer';
import {
  Expression,
  IfNode,
  IfElseNode,
  AndNode,
  OrNode,
  ParensNode,
  EqualsNode,
  NotEqualsNode,
  ForEachNode,
  StringNode,
  IntNode,
  NullNode,
  ReferenceNode,
  QuietReferenceNode,
  ObjectNode,
  ListNode,
  FloatNode,
  QuotesNode,
  RawNode,
  SetNode,
  CompoundExpressionNode,
  CommentNode,
  ToJsonNode,
  BooleanNode,
  NotNode,
  NewLineNode,
  ReturnNode,
  parens,
  IsNullOrEmptyNode,
} from '../ast';

const TAB = '  ';

class VtlPrinter extends Printer {
  protected printIf(node: IfNode, indent: string = '') {
    if (node.inline) {
      return `#if( ${this.printExpr(node.predicate, '')} ) ${this.printExpr(node.expr, '')} #end`;
    }
    return `${indent}#if( ${this.printExpr(node.predicate, '')} )\n${this.printExpr(node.expr, indent + TAB)}\n${indent}#end`;
  }

  protected printIfElse(node: IfElseNode, indent: string = '') {
    if (node.inline) {
      return (
        `#if( ${this.printExpr(node.predicate)} ) ` +
        `${this.printExpr(node.ifExpr)} ` +
        `#else ` +
        `${this.printExpr(node.elseExpr)} ` +
        `#end`
      );
    }
    return (
      `${indent}#if( ${this.printExpr(node.predicate)} )\n` +
      `${this.printExpr(node.ifExpr, indent + TAB)}\n` +
      `${indent}#else\n` +
      `${this.printExpr(node.elseExpr, indent + TAB)}\n` +
      `${indent}#end`
    );
  }

  protected printAnd(node: AndNode, indent: string = ''): string {
    return indent + node.expressions.map((e: Expression) => this.printExpr(e)).join(' && ');
  }

  protected printOr(node: OrNode, indent: string = ''): string {
    return indent + node.expressions.map((e: Expression) => this.printExpr(e)).join(' || ');
  }

  protected printParens(node: ParensNode, indent: string = ''): string {
    return `${indent}(${this.printExpr(node.expr)})`;
  }

  protected printEquals(node: EqualsNode, indent: string = ''): string {
    return `${indent}${this.printExpr(node.leftExpr)} == ${this.printExpr(node.rightExpr)}`;
  }

  protected printNotEquals(node: NotEqualsNode, indent: string = ''): string {
    return `${indent}${this.printExpr(node.leftExpr)} != ${this.printExpr(node.rightExpr)}`;
  }

  protected printForEach(node: ForEachNode, indent: string = ''): string {
    return (
      `${indent}#foreach( ${this.printExpr(node.key)} in ${this.printExpr(node.collection)} )\n` +
      node.expressions.map((e: Expression) => this.printExpr(e, indent + TAB)).join('\n') +
      `\n${indent}#end`
    );
  }

  protected printString(node: StringNode): string {
    return `"${node.value}"`;
  }

  protected printBool(node: BooleanNode): string {
    return `${node.value}`;
  }

  protected printRaw(node: RawNode, indent: string = ''): string {
    return `${indent}${node.value}`;
  }

  protected printQuotes(node: QuotesNode): string {
    return `"${this.printExpr(node.expr)}"`;
  }

  protected printInt(node: IntNode): string {
    return `${node.value}`;
  }

  protected printFloat(node: FloatNode): string {
    return `${node.value}`;
  }

  protected printNull(node: NullNode): string {
    return `null`;
  }

  protected printReference(node: ReferenceNode): string {
    return `\$${node.value}`;
  }

  protected printQuietReference(node: QuietReferenceNode, indent: string = ''): string {
    const val = typeof node.value === 'string' ? node.value : this.printExpr(node.value);
    return `${indent}$util.qr(${val})`;
  }

  printObject(node: ObjectNode, indent: string = ''): string {
    const attributes = node.attributes.map((attr: [string, Expression], i: number) => {
      return `${indent}${TAB}"${attr[0]}": ${this.printExpr(attr[1], indent + TAB)}${i < node.attributes.length - 1 ? ',' : ''}`;
    });
    const divider = attributes.length > 0 ? `\n${indent}` : '';
    return `{${divider}${attributes.join(divider)}${divider}}`;
  }

  protected printList(node: ListNode, indent: string = ''): string {
    const values = node.expressions.map((e: Expression) => this.printExpr(e, '')).join(', ');
    return `${indent}[${values}]`;
  }

  protected printSet(node: SetNode, indent: string = ''): string {
    return `${indent}#set( ${this.printReference(node.key)} = ${this.printExpr(node.value, '')} )`;
  }

  protected printComment(node: CommentNode, indent: string = ''): string {
    return `${indent}## ${node.text} **`;
  }

  protected printCompoundExpression(node: CompoundExpressionNode, indent: string = ''): string {
    if (node.recurseIndent) {
      return node.expressions.map((node: Expression) => this.printExpr(node, indent)).join(node.joiner);
    }
    return indent + node.expressions.map((node: Expression) => this.printExpr(node)).join(node.joiner);
  }

  protected printToJson(node: ToJsonNode, indent: string = ''): string {
    return `${indent}$util.toJson(${this.printExpr(node.expr, '')})`;
  }

  protected printIsNullOrEmpty(node: IsNullOrEmptyNode, indent: string = ''): string {
    return `${indent}$util.isNullOrEmpty(${this.printExpr(node.expr, '')})`;
  }

  protected printNot(node: NotNode, indent: string = ''): string {
    return `${indent}!${this.printExpr(node.expr, '')}`;
  }

  protected printNewLine(node: NewLineNode): string {
    return '\n';
  }

  protected printReturn(node: ReturnNode, indent: string = ''): string {
    let suffix: string = '';
    if (node.value !== undefined) {
      suffix = this.printParens(parens(node.value));
    }
    return `${indent}#return` + suffix;
  }
}

export const vtlPrinter = new VtlPrinter();
