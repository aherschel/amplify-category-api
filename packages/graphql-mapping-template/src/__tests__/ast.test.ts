import { ref, obj, str, forEach, qref, set, compoundExpression, ifElse, nul, bool, methodCall, list, Expression } from '../ast';
import { DynamoDBMappingTemplate } from '../dynamodb';
import { vtlPrinter, jsPrinter } from '../printer';

const validateResolver = (expression: Expression): void => {
  expect(vtlPrinter.print(expression)).toMatchSnapshot();
  expect(jsPrinter.print(expression)).toMatchSnapshot();
};

test('create a put item resolver with the ast', () => {
  validateResolver(
    DynamoDBMappingTemplate.putItem({
      key: obj({
        type: str('Post'),
        id: ref('util.autoId()'),
      }),
      attributeValues: obj({
        value: methodCall(ref('util.dynamodb.toMapJson'), ref('ctx.input')),
      }),
    }),
  );
});

test('create a query resolver with the ast', () => {
  validateResolver(
    DynamoDBMappingTemplate.query({
      query: obj({
        expression: str('#typename = :typename'),
        expressionNames: obj({
          '#typename': str('__typename'),
        }),
        expressionValues: obj({
          ':typename': obj({
            S: str('test'),
          }),
        }),
      }),
      scanIndexForward: bool(true),
      filter: ifElse(
        ref('context.args.filter'),
        methodCall(ref('util.transform.toDynamoDBFilterExpression'), ref('ctx.args.filter')),
        nul(),
      ),
      limit: ref('limit'),
      nextToken: ifElse(ref('context.args.nextToken'), ref('context.args.nextToken'), nul()),
    }),
  );
});

test('create a response mapping template that merges a nested object', () => {
  validateResolver(
    compoundExpression([
      set(ref('result'), methodCall(ref('util.map.copyAndRemoveAllKeys'), ref('context.result'), list([str('value')]))),
      forEach(ref('entry'), ref('context.result.value.entrySet()'), [
        qref(methodCall(ref('result.put'), ref('entry.key'), ref('entry.value'))),
      ]),
      methodCall(ref('util.toJson'), ref('result')),
    ]),
  );
});
