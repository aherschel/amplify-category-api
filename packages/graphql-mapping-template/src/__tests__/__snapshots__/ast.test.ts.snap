// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create a put item resolver with the ast 1`] = `
"{
  \\"version\\": \\"2018-05-29\\",
  \\"operation\\": \\"PutItem\\",
  \\"key\\": {
      \\"type\\": \\"Post\\",
      \\"id\\": $util.autoId()
  },
  \\"attributeValues\\": {
      \\"value\\":     $util.dynamodb.toMapJson($ctx.input)
  },
  \\"condition\\": 
}"
`;

exports[`create a put item resolver with the ast 2`] = `
"{
  \\"version\\": \\"2018-05-29\\",
  \\"operation\\": \\"PutItem\\",
  \\"key\\": {
      \\"type\\": \\"Post\\",
      \\"id\\": util.autoId()
  },
  \\"attributeValues\\": {
      \\"value\\":     util.dynamodb.toMapJson(ctx.input)
  },
  \\"condition\\": 
}"
`;

exports[`create a query resolver with the ast 1`] = `
"{
  \\"version\\": \\"2018-05-29\\",
  \\"operation\\": \\"Query\\",
  \\"query\\": {
      \\"expression\\": \\"#typename = :typename\\",
      \\"expressionNames\\": {
          \\"#typename\\": \\"__typename\\"
    },
      \\"expressionValues\\": {
          \\":typename\\": {
              \\"S\\": \\"test\\"
      }
    }
  },
  \\"scanIndexForward\\": true,
  \\"filter\\":   #if( $context.args.filter )
    $util.transform.toDynamoDBFilterExpression($ctx.args.filter)
  #else
null
  #end,
  \\"limit\\": $limit,
  \\"nextToken\\":   #if( $context.args.nextToken )
$context.args.nextToken
  #else
null
  #end
}"
`;

exports[`create a query resolver with the ast 2`] = `
"{
  \\"version\\": \\"2018-05-29\\",
  \\"operation\\": \\"Query\\",
  \\"query\\": {
      \\"expression\\": \\"#typename = :typename\\",
      \\"expressionNames\\": {
          \\"#typename\\": \\"__typename\\"
    },
      \\"expressionValues\\": {
          \\":typename\\": {
              \\"S\\": \\"test\\"
      }
    }
  },
  \\"scanIndexForward\\": true,
  \\"filter\\":   if( context.args.filter ) {
    util.transform.toDynamoDBFilterExpression(ctx.args.filter)
  } else {
null
  },
  \\"limit\\": limit,
  \\"nextToken\\":   if( context.args.nextToken ) {
context.args.nextToken
  } else {
null
  }
}"
`;

exports[`create a response mapping template that merges a nested object 1`] = `
"#set( $result = $util.map.copyAndRemoveAllKeys($context.result, [\\"value\\"]) )
#foreach( $entry in $context.result.value.entrySet() )
  $util.qr($result.put($entry.key, $entry.value))
#end
$util.toJson($result)"
`;

exports[`create a response mapping template that merges a nested object 2`] = `
"result = util.map.copyAndRemoveAllKeys(context.result, [\\"value\\"])
for ( const entry in context.result.value.entrySet() ) {
  result.put(entry.key, entry.value)
}
util.toJson(result)"
`;
